type_ck integration

jvanasco opened the pull request:
This is the first attempt at #4784 and #4790. I expect requested changes on the error messages at the very least, but wanted to post it for review.

### Description

* type_ck has been added
* `_prefix_dict` was renamed `_base_prefix_dict` along with a note to better describe how prefix overrides work (such as 'ck' -> 'type_ck')
* a small change to `_get_prefixes`: only use the `type_` prefix if no name was presented. this was more backwards compatible after writing some test cases.
* test cases! test_metadata has a lot of rewrites and additions to NamingConventionsTest , starting with `test_schematype_ck_name_boolean` and all the way down to `test_schematype_no_ck_name_boolean_no_name`
* newly added- `ConventionDict.debug_info(`
* handled edgecase & test for a constraint with no columns, but a column_name token. improved error message, also added  "debug_info".
* improved error message for having convention that requires a constraint name, but no constraint name available; also added  "debug_info".

### Checklist
This pull request is:
- [x] A documentation / typographical error fix
- [x] A short code fix
        - Fixes: #4784
        - Fixes: #4790
- [ ] A new feature implementation

### Of Importance...

I added a docstring to each test describing the use-case.  These were either explicit from the various tickets or inferred by my best guesses.  They need review.

I also added 2 tests for dealing with PEP-435 enums. The sqlalchemy docs note the classname will be used as the constraint_name, however there were no previous tests concerning this.

jvanasco committed:
work on #4784 regarding constraints
* introduced new type_ck prefix for type-bound constraints; code and docs
* misc unit tests to cover the type-bound items
* added unit tests to ensure pep-compliant enum objects are used to generate a name as the docs instruct

work on #4790 regarding exception text
* added some debug info to constraint naming exceptions; this is likely to change

Change-Id: I48dd77301c10d6462847e2d00ac1412be20a74b2

zzzeek commented on doc/build/core/constraints.rst:
```diff
@@ -697,20 +703,22 @@ normally by using a convention which includes ``%(constraint_name)s``
 
     metadata = MetaData(
-        naming_convention={"ck": "ck_%(table_name)s_%(constraint_name)s"}
+        naming_convention={"type_ck": "type_ck_%(table_name)s_%(constraint_name)s"}
```
I think I might want to keep the rendered example prefix as "ck_", because, when database identifiers get too long, things start getting weird.   at this point we have good truncation logic for these names but especially on a backend like Oracle names are truncated very short.

jvanasco replied:
ok.

zzzeek commented on doc/build/core/constraints.rst:
```diff
@@ -725,20 +733,32 @@ which works nicely with :class:`.SchemaType` since these constraints have
 
     metadata = MetaData(
-        naming_convention={"ck": "ck_%(table_name)s_%(column_0_name)s"}
+        naming_convention={"ck": "ck_%(table_name)s_%(constraint_name)s",
```
I'd change the code formatting here, if you want, we can just pull this into gerrit and I can push additional tweaks and whatnot.

jvanasco replied:
let me handle most of this in github for a bit first. i still have issues with gerrit and I'll lose too much time on getting up-to-speed.

zzzeek commented on test/sql/test_metadata.py:
```diff
@@ -4755,16 +4792,52 @@ def test_schematype_ck_name_boolean_not_on_name(self):
+        # but is hit at compile time
+        assert_raises_message(
+            exc.InvalidRequestError,
+            r"Naming convention including \%\(constraint_name\)s token "
```
OK so, more monkey wrenches, when the user gets this error, the thing they likely *want* to do is add a "type_ck" convention to their dictionary, because now they can!   So the message should refer to that.   Since this is completely confusing, you an also add a "code" to the exception to point them to a doc blurb at https://docs.sqlalchemy.org/en/13/errors.html that explains the whole story.     this one is a real quagmire.

jvanasco replied:
will do!

zzzeek commented on test/sql/test_metadata.py:
```diff
@@ -4755,16 +4792,52 @@ def test_schematype_ck_name_boolean_not_on_name(self):
+            exc.InvalidRequestError,
+            r"Naming convention including \%\(constraint_name\)s token "
+            r"requires that constraint is explicitly named."
+            r" Constraint Identifiers: Table- `user`; Column\(s\)- `x`.",
```
also I'd put the "table" / "column" info more in sentence format above , "naming convention "ck_%(foo)s_%(bar)s" failed to apply to CHECK constraint on table X with column Q since the constraint has no name, and the convention uses the "%(constraint_name)s" token.     Consider using the "type_ck" naming convention for type-bound CHECK constraints so that no name is necessary.   more inforamtion at https://sqlalche.me/foobar"

jvanasco replied:
will do!

jvanasco replied:
@zzzeek where are the short urls maintained?

zzzeek replied:
you make up a random four letter code, you add it as "code='ABCD'" to the InvalidRequestError() or whatever it is, then you write it in https://github.com/sqlalchemy/sqlalchemy/blob/master/doc/build/errors.rst

zzzeek commented on test/sql/test_metadata.py:
```diff
@@ -4755,16 +4792,52 @@ def test_schematype_ck_name_boolean_not_on_name(self):
+        """If a CheckConstraint is EXPLICITLY declared for a Enum column, it
+        should use the `ck_` prefix and not the `type_ck`."""
+        m1 = MetaData(
+            naming_convention={"ck": "ck_%(table_name)s_%(constraint_name)s",
```
have you tried out the [pre-commit](https://pre-commit.com/) tool?   Like everything, took me awhile to be open towards it but all of my commits are auto-black formatted now, which is critical because otherwise they get auto-formatted in a later commit and make a mess.

jvanasco replied:
i can git it a shot. if not, i'll just run black manually.

zzzeek commented on test/sql/test_metadata.py:
```diff
@@ -4780,11 +4853,190 @@ def test_schematype_ck_name_enum(self):
+            [c for c in u1.constraints if isinstance(c, CheckConstraint)][
+                0
+            ].name,
+            "_unnamed_",
```
I hate the "_unnamed_" thing, can we maybe just do an all 1.4 commit here that whacks that whole "_defer_none_name" thing (assuming this is possible, I want to have a default "type_ck" convention to help with this) ?   I can work on that since it's my mess to clean up.

jvanasco replied:
1. I dislike the unnamed thing too.
2. I'm not sure `_defer_none_name` can be dropped.  if it can, I think it will be a lot of work and break many test cases.  a relevant test-case is the new one I added for your edge-case of columnless constraints -- `[test_schematype_ck_no_columns_on_constraint](https://github.com/sqlalchemy/sqlalchemy/pull/4792/files/6182721f8556a08485f1b089f24cf30f4705d1ed#diff-85c40f646c916a1455f90ca632061d28R5006-R5035)` (you may need to scroll all the way to the bottom; github is highlighting the code but not autoscrolling to the section)

recreated here...

            u1 = Table(
                "user", m1,
                Column("x", Integer),
                CheckConstraint('z > 10')
            )

This will automatically raise the error on `Table` - instead of `.compile` like the other cases - because the constraint name doesn't exist but it also isn't deferred.

zzzeek reviewed (commented):
I put some thoughts in but I havent gone through all of it yet.     this is a hard one because the naming thing is so weird.

jvanasco commented:
integrated the changes and updated some more error messages.

i didn't run black on a few files (notably tests, but IIRC also schema) because it was clear black had not been run on them before, and the actual changes would have been hidden.

one of the bigger changes in this, is tracking the "active convention", which is really an active naming convention template.  this is required for the improved error messages.

LouisTrezzini commented:
This looks great! Do you have any plans to have this feature merged in an upcoming release?

zzzeek commented:
someday... :)   I really have a backlog

CaselIT committed:
Update select usage to use the new 1.4 format

This change includes mainly that the bracketed use within
select() is moved to positional, and keyword arguments are
removed from calls to the select() function.  it does not
yet fully address other issues such as keyword arguments passed
to the table.select().

Additionally, allows False / None to both be considered
as "disable" for all of select.correlate(), select.correlate_except(),
query.correlate(), which establishes consistency with
passing of ``False`` for the legact select(correlate=False)
argument.

Change-Id: Ie6c6e6abfbd3d75d4c8de504c0cf0159e6999108

CaselIT committed:
Fix typo in :paramref:`_engine.create_engine.isolation_level` names

Fixes: #5563
Change-Id: I29204fdf679d750c66ed17daf70bc8d7cb1b7f65

zzzeek committed:
Repair documentation issues

Fixes some remaining issues detailed at #5428.

Fixes: #5428
Change-Id: I942a64411766fc82f30791eee570747a218af77d

zzzeek committed:
Add more docs for populate_existing(); link with_for_update

The populate_existing() method is actually changing
to be an execution option, however it has almost no
mention in the narrative docs so add docs in terms of the
1.x version first, including that we mention you almost
definitely want to use this method if you are also using
with_for_update().

Fixes: #5572
Fixes: #4774
Change-Id: Ieca916400622c1ffc1ae81204132a02a0983594c

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision e407db8c3b59e93d5823f224a05ab0f32a463b5b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Failed to create a gerrit review, git squash against branch 'master' failed

zzzeek commented:
hi @jvanasco , do a rebase on this and I can pull it into gerrit.

jvanasco commented:
Closing this because the merging/etc is a nightmare. I will generate a new PR and pull it into gerrit.
